### ISSUE TITLE
Error with psycopg2-binary while building docker version

### ISSUE THREAD

**Issue by MaxValue** [rr-/szurubooru#192]
So when running `docker-compose build --pull` according to the installation guide, it throws this error:

> Collecting psycopg2-binary>=2.6.1 (from -r ./requirements.txt (line 3))
>   Downloading https://files.pythonhosted.org/packages/17/dc/3f6bfae668594013cb9ef0cba4f4863451220944650868ad67332ca9069c/psycopg2-binary-2.7.5.tar.gz (427kB)
>     Complete output from command python setup.py egg_info:
>     running egg_info
>     creating pip-egg-info/psycopg2_binary.egg-info
>     writing pip-egg-info/psycopg2_binary.egg-info/PKG-INFO
>     writing dependency_links to pip-egg-info/psycopg2_binary.egg-info/dependency_links.txt
>     writing top-level names to pip-egg-info/psycopg2_binary.egg-info/top_level.txt
>     writing manifest file 'pip-egg-info/psycopg2_binary.egg-info/SOURCES.txt'
>     
>     Error: pg_config executable not found.
>     
>     pg_config is required to build psycopg2 from source.  Please add the directory
>     containing pg_config to the $PATH or specify the full executable path with the
>     option:
>     
>         python setup.py build_ext --pg-config /path/to/pg_config build ...
>     
>     or with the pg_config option in 'setup.cfg'.
>     
>     If you prefer to avoid building psycopg2 from source, please install the PyPI
>     'psycopg2-binary' package instead.
>     
>     For further information please check the 'doc/src/install.rst' file (also at
>     <http://initd.org/psycopg/docs/install.html>).
>     
>     
>     ----------------------------------------
> Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-v13guz68/psycopg2-binary/
> ERROR: Service 'backend' failed to build: The command '/bin/sh -c pip3 install --no-cache-dir -r ./requirements.txt' returned a non-zero code: 1

Is the bug in the psycopg2-binary module or can we simply configure the building scripts differently?

FYI: This error occured on Rasbian stretch but not on Ubuntu xenial.

**Comment by joshuaavalon** [comment]
> This error occured on Rasbian stretch but not on Ubuntu xenial.

You changed the base image?

**Comment by MaxValue** [comment]
Nope I did not change anything about the docker image. These are the complete installation steps I did:

#### Ubuntu
##### Docker
I followed [this tutorial](https://docs.docker.com/install/linux/docker-ce/ubuntu/) except I ignored [the section about the storage drivers](https://docs.docker.com/install/linux/docker-ce/ubuntu/#supported-storage-drivers). I then installed docker via `sudo apt-get install docker-ce`.
##### Szurubooru
I followed [the installation manual](https://github.com/rr-/szurubooru/blob/master/INSTALL.md) and also increased the `vm.max_map_count` to 262144 according to [this manual for linux](https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html#docker-cli-run-prod-mode).

#### Raspbian
##### Docker
I used [this convenience script](https://docs.docker.com/install/linux/docker-ce/debian/#install-using-the-convenience-script) for installing docker on raspbian and installed `docker-compose` via pip: `python3 -m pip install docker-compose`
##### Szurubooru
I followed [the installation manual](https://github.com/rr-/szurubooru/blob/master/INSTALL.md) and also increased the `vm.max_map_count` to 262144 according to [this manual for linux](https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html#docker-cli-run-prod-mode).

So to sum it up, the only thing which differed between installs is the way docker itself got installed.

**Comment by joshuaavalon** [comment]
My guess would be your raspberry pi CPU. [psycopg2](https://pypi.org/project/psycopg2/#files) does not have a compiled version for the CPU and it try to compile and failed because of missing headers. Others does not need this because they can just download the compiled package.

You can add back the necessary headers before this line
https://github.com/rr-/szurubooru/blob/b22c887e4b7ebc3c11d72f2ce9e8086403616c46/server/Dockerfile#L38

**Comment by MaxValue** [comment]
I will try that! Thanks!
I will post the results here, once I tested it.

**Comment by WychWitch** [comment]
Hello, I've run into this problem as well! How do you add back the necessary headers? I'm also using a a raspberry pi (ARM)

**Comment by MaxValue** [comment]
No problem!

I tried it once (meaning: compiling), but it threw just more errors which were even more intelligible (at least for me) so I kinda gave up. Although I still would like to see this solved.
I guess you have to keep installing all dependencies, google all the errors and ... I don know. :man_shrugging:
I never really had to compile anything in my life, so this is very black magic to me.

**Comment by WychWitch** [comment]
With the help of a friend I solved it! But now I ran into another error. I'll post the fixed dockerfile in here once I get everything smoothed out!

**Comment by SirFireNewt** [comment]
anyone have a workaround for this issue?

**Comment by joshuaavalon** [comment]
@SirFireNewt You need to find what dependencies is missing on your host OS or you a  host OS that support by psycopg2.

**Comment by MaxValue** [comment]
Since there has been interest for this again, @WychWitch, please just post what you have. It doesn't have to be perfect. A few pointers in the right direction are much more helpful than nothing :)

**Comment by WychWitch** [comment]
Ah sorry I meant to post on here sooner! It's getting late so I'll grab my files and post them tomorrow!!

I gotta warn you though, I never got this working on my rpi. I gave up and bought an rpi3 and installed it on there instead (though it still needed some customization)

**Comment by WychWitch** [comment]
Sorry for the radio silence! I got hella busy and completely forgot until now!

Here are my files! If this isn't enough let me know, I might've forgotten one. Again, this is only for rpi3/arm32v7

replace all of the [name] and [password] with your own!

**Client/Dockerfile**
```Docker
FROM arm32v7/node:9 as builder
WORKDIR /opt/app

COPY package.json ./
RUN npm install

COPY . ./

ARG BUILD_INFO="docker-latest"
ARG CLIENT_BUILD_ARGS=""
RUN BASE_URL="__BASEURL__" node build.js --gzip ${CLIENT_BUILD_ARGS}


FROM tobi312/rpi-nginx:alpine
WORKDIR /var/www

RUN \
    # Create init file
    echo "#!/bin/sh" >> /init && \
    echo 'sed -i "s|__BACKEND__|${BACKEND_HOST}|" /etc/nginx/nginx.conf' >> /init && \
    echo 'sed -i "s|__BASEURL__|${BASE_URL:-/}|g" /var/www/index.htm /var/www/manifest.json' >> /init && \
    echo 'exec nginx' >> /init && \
    chmod a+x /init

CMD ["/init"]
VOLUME ["/data"]

COPY nginx.conf.docker /etc/nginx/nginx.conf
COPY --from=builder /opt/app/public/ .
```
**Server/Dockerfile**
```
FROM scratch as approot
WORKDIR /opt/app

COPY alembic.ini wait-for-es generate-thumb ./
COPY szurubooru/ ./szurubooru/
COPY config.yaml.dist ./


FROM python:3.6-slim
WORKDIR /opt/app

ARG PUID=1000
ARG PGID=1000
ARG PORT=6666
RUN \
    # Set users
    mkdir -p /opt/app /data && \
    groupadd -g ${PGID} app && \
    useradd -d /opt/app -M -c '' -g app -r -u ${PUID} app && \
    chown -R app:app /opt/app /data && \
    # Create init file
    echo "#!/bin/sh" >> /init && \
    echo "set -e" >> /init && \
    echo "cd /opt/app" >> /init && \
    echo "./wait-for-es" >> /init && \
    echo "alembic upgrade head" >> /init && \
    echo "exec waitress-serve --port ${PORT} szurubooru.facade:app" \
        >> /init && \
    chmod a+x /init && \
    # Install ffmpeg
    apt-get -yqq update && \
    apt-get -yq install --no-install-recommends ffmpeg libpq-dev make gcc libffi-dev zlib1g-dev libjpeg-dev libtiff-dev libfreetype6-dev liblcms2-dev libwebp-dev tcl-dev && \
    rm -rf /var/lib/apt/lists/* && \
    # Install waitress
    pip3 install --no-cache-dir waitress

COPY --chown=app:app requirements.txt ./requirements.txt
RUN pip3 install --no-cache-dir -r ./requirements.txt

# done to minimize number of layers in final image
COPY --chown=app:app --from=approot / /

VOLUME ["/data/"]
EXPOSE ${PORT}
USER app
CMD ["/init"]
```
**.env**
```
# Database credentials
POSTGRES_USER=[name]
POSTGRES_PASSWORD=[password]

# This shows up on the homescreen, indicating build information
BUILD_INFO=latest

# Port to expose HTTP service
PORT=[port]

# Directory to store image data
## remember to chmod this directory
MOUNT_DATA=/var/local/[name]/data

# Directory to store database files
MOUNT_SQL=/var/local/[name]/sql
```

**requirements.txt**
```
alembic>=0.8.5
pyyaml>=3.11
psycopg2-binary>=2.6.1
SQLAlchemy>=1.0.12
coloredlogs==5.0
elasticsearch>=5.0.0
elasticsearch-dsl>=5.0.0
numpy>=1.8.2
pillow>=4.3.0
pynacl==1.2.1
pytz>=2018.3
pyRFC3339>=1.0
```

**docker-compose.yml**;
```
## Example Docker Compose configuration
##
## Use this as a template to set up docker-compose, or as guide to set up other
## orchestration services
version: '2'

services:
  api:
    build:
      context: ./server
    depends_on:
      - sql
      - elasticsearch
    environment:
      ## These should be the names of the dependent containers listed above,
      ## or FQDNs/IP addresses if these services are running outside of Docker
      POSTGRES_HOST: sql
      ESEARCH_HOST: elasticsearch
      ## Credentials for database:
      POSTGRES_USER: [Name]
      POSTGRES_PASSWORD: [pass]
      ## Commented Values are Default:
      #POSTGRES_DB: defaults to same as POSTGRES_USER
      #POSTGRES_PORT: 5432
      #ESEARCH_PORT: 9200
      #ESEARCH_INDEX: szurubooru
      #LOG_SQL: 0 (1 for verbose SQL logs)
    volumes:
      - "${MOUNT_DATA}:/data"
      - "./server/config.yaml:/opt/app/config.yaml"
  client:
    build:
      context: ./client
      args:
        BUILD_INFO:
    depends_on:
      - api
    environment:
      BACKEND_HOST: api
    volumes:
      - "${MOUNT_DATA}:/data:ro"
    ports:
      - "${PORT}:80"
  sql:
    image: postgres:alpine
    restart: unless-stopped
    environment:
      POSTGRES_USER: [name]
      POSTGRES_PASSWORD: [pass]
    volumes:
      - "${MOUNT_SQL}:/var/lib/postgresql/data"
  elasticsearch:
    image: ind3x/rpi-elasticsearch
    environment:
      ## Specifies the Java heap size used
      ## Read
      ##  https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html
      ## for more info
      ES_JAVA_OPTS: -Xms512m -Xmx512m
    volumes:
      - index:/usr/share/elasticsearch/data

volumes:
  index: # Scratch space for ElasticSearch index, will be rebuilt if lost
```
